include/keisan/hungarian.impl.hpp

```diff
@@ -0,0 +1,372 @@
+ * For each row and column, find the lowest element and
+ * subtract it from each element in that row or column. */
+template<size_t N>
+void Hungarian<N>::step_1()
+{
+  // Subtract row minima
+  for (int row = 0; row < actual_size; ++row) {
+    double lowest_value = matrix[row][0];
+    for (int col = 1; col < actual_size; ++col) {
+      if (matrix[row][col] < lowest_value) {
+        lowest_value = matrix[row][col];
+      }
+    }
+
+    for (int col = 0; col < actual_size; ++col) {
+      matrix[row][col] -= lowest_value;
+    }
+  }
+
+  // Subtract column minima
+  for (int col = 0; col < actual_size; ++col) {
+    double lowest_value = matrix[0][col];
+    for (int row = 1; row < actual_size; ++row) {
+      if (matrix[row][col] < lowest_value) {
+        lowest_value = matrix[row][col];
+      }
+    }
+
+    for (int row = 0; row < actual_size; ++row) {
+      matrix[row][col] -= lowest_value;
+    }
+  }
+
+  step = 2;
+}
+
+/* Step 2: Star the first zero
+ * Find a zero in the matrix. If there is no starred zero in its row or column,
+ * star it and cover its row and column. Repeat until there are no uncovered
+ * zeros left. */
+template<size_t N>
+void Hungarian<N>::step_2()
+{
+  for (int row = 0; row < actual_size; ++row) {
+    for (int col = 0; col < actual_size; ++col) {
+      if (matrix[row][col] == 0 && row_cover[row] == UNCOVERED && col_cover[col] == UNCOVERED) {
+        mask[row][col] = STAR;
+        row_cover[row] = COVERED;
+        col_cover[col] = COVERED;
+      }
+    }
+  }
+
+  clear_covers(row_cover);
+  clear_covers(col_cover);
+
+  step = 3;
+}
+
+/* Step 3: Cover columns with starred zeros
+ * Cover every column that has a starred zero in it. If K columns are covered,
+ * the starred zeros describe a complete matching, else go to step 4. */
+template<size_t N>
+void Hungarian<N>::step_3()
+{
+  int col_count = 0;
+  for (int row = 0; row < actual_size; ++row) {
+    for (int col = 0; col < actual_size; ++col) {
+      if (mask[row][col] == STAR && col_cover[col] == UNCOVERED) {
+        col_cover[col] = COVERED;
+        col_count++;
+      }
+    }
+  }
+
+  if (col_count >= actual_size) {
+    step = 7;
+  } else {
+    step = 4;
+  }
+}
+
+/* Step 4: Create a star zero
+ * Find a non-covered zero and prime it. If there is no starred zero in its row,
+ * go to step 5. Otherwise, cover the row and uncover the column of the starred
+ * zero. Repeat until there are no uncovered zeros left. */
+template<size_t N>
+void Hungarian<N>::step_4()
+{
+  int row, col;
+  while (true) {
+    find_a_zero(row, col);
+    if (row == -1 || col == -1) {
+      step = 6;
+      return;
+    } else {
+      mask[row][col] = PRIME;
+      if (star_in_row(row)) {
+        col = find_star_in_row(row);
+        row_cover[row] = COVERED;
+        col_cover[col] = UNCOVERED;
+      } else {
+        path_row_0 = row;
+        path_col_0 = col;
+        step = 5;
+        return;
+      }
+    }
```

Comment:
The ```else``` block can be removed because the ```if``` condition returns the function.
```suggestion
    if (row == -1 || col == -1) {
      step = 6;
      return;
    }

    mask[row][col] = PRIME;
    if (!star_in_row(row)) {
        path_row_0 = row;
        path_col_0 = col;
        step = 5;
        return;
     }

      col = find_star_in_row(row);
      row_cover[row] = COVERED;
      col_cover[col] = UNCOVERED;
```